### ISSUE TITLE
Proposal: Metrics for log driver log loss, logs sent, throughput, in Docker Stats

### ISSUE THREAD

**Issue by PettitWesley** [moby/moby#45953]
### Description

Recently, I worked on a project to benchmark log loss with the AWSLogs driver in `non-blocking` mode with different values of `max-buffer-size` and log output throughput. These results will be publishes soon and will be linked here once published. 

One of the issues I see with users using the benchmark results to guide their own configurations is that there’s limited visibility into logging with log drivers. I can’t easily find the output rate of my container. When the [buffer](https://github.com/moby/moby/blob/master/daemon/logger/ring.go#L160) drops logs for lack of buffer space, there’s no log message or metric emitted. 

### Problem Statement

Log drivers have limited visibility. Ideally, users should be able to track metrics or logs on the rate of log emission from their containers and the rate of log loss vs log upload success from the driver. 

With `non-blocking` mode, the main concern is log loss in the [buffer](https://github.com/moby/moby/blob/master/daemon/logger/ring.go#L160). When logs are dropped, it is completely silent, there’s no way to track it. 

The highest priority problem is lack of visibility into log loss. Tracking log throughput is secondary. 

### Proposal: Metrics for log driver log loss, logs sent, throughput, in Docker Stats

While the log loss visibility problem could be solved with logs emitted by the Docker daemon, this approach doesn’t scale. Users want to quantify log loss per container, and counting messages emitted by the entire daemon is inconvenient. Also, in cases of repeated high rate of log loss, the docker daemon logs would be spammed with many error statements.

The ideal place for this sort of data is to add new counters in the `docker stats` interface.

**Comment by thaJeztah** [comment]
I'm a bit on the fence on making this part of `docker stats`, but I know we did discuss making "drops" more visible. Some very quick "off the bat" thoughts;

- wondering if this is a counter we could add to the prometheus metrics
- I think we discussed adding a log-entry to the local logs buffer (which is used to read back last X logs when using `docker logs`), perhaps above a certain threshold.
- I _guess_ some of this would depend on the driver that's used (detecting dropped messages?)


/cc @corhere @cpuguy83 any thoughts?

**Comment by corhere** [comment]
I think that Prometheus metrics would be ideal. We could label the metrics with all sorts of dimensions (e.g. log driver, container id) so that users can slice, dice and aggregate in different ways. Users would get time-series data with a whole ecosystem to ingest and analyze it, and all we'd have to do is increment some event counters.

On the other hand, `docker stats` has a very low barrier to entry for users. Exposing some basic per-container log stats over that interface could help people run experiments and troubleshoot interactively without first having to provision and configure a Prometheus stack. But that takes both Engine API and CLI work to implement as the stats API data model [is not exactly extensible.](https://github.com/moby/moby/blob/193f16238bb5ce1eb87efd8da03dce41466ef8c4/api/types/stats.go)

> * I _guess_ some of this would depend on the driver that's used (detecting dropped messages?)

The drivers couldn't even detect dropped messages if they wanted to! The `RingLogger` which is inserted in front of the logging driver drops log messages when the buffer fills up, and (IIRC) the log driver is not informed when that happens. Though I'm sure that there could be plenty of driver-specific metrics that they may want to export, possibly including counters for log messages dropped _by the driver_. But that's really secondary to the main focus of this proposal.

**Comment by PettitWesley** [comment]
I edited my proposal to add a nice diagram I made showing now the non-blocking buffer comes before the log driver and thus that one could be tracked easily without the need to change every single log driver.

**Comment by createchange** [comment]
Hi - it looks like this issue may have fizzled out, but I am coming by way of Trusted Advisor checks and, subsequently, this article: https://aws.amazon.com/blogs/containers/preventing-log-loss-with-non-blocking-mode-in-the-awslogs-container-log-driver/

I work in Security for a healthcare entity, and am trying to strike a balance for my Engineering team. I would like to permit the Engineering team to configure the log driver to be in non-blocking mode, which could alleviate situations where real time healthcare information is halted from ingest/processing. I don't know that we have ever had a service actually halted due to inability to ship logs to CloudWatch, but it is now on my radar and something that I want to account for.

Unfortunately, without the ability to tell if the buffer is full and dropping logs, it feels rather unsafe to permit them to move the drivers into non-blocking mode, as we can't monitor for if/when logs are being dropped, which would signify that the buffer level is set too low. As such, we are left facing the potential that patient data processing is halted because the driver must remain in blocking mode.

With all that in mind, I would very much so appreciate having an event emitted to CloudWatch which can be captured and alerted on when logs are being dropped, which would indicate the need to increase the buffer size. Until monitoring on dropped logs is implemented, the non-blocking functionality is not something that I feel comfortable socializing and, if I can, is likely something that I will be putting an SCP in place around to prevent this, as - of course - in a healthcare situation, logs are of utmost importance for compliance and regulatory purposes.

On a related note - in the aforementioned article, I see great guidance on how to deduce what an appropriate buffer size is for general use cases, however I don't see any drawbacks enumerated about setting a high buffer size. Are there any performance drawbacks to setting a buffer size that is 25mb? 50mb? Is the main drawback that this could fill up more memory, and thus increase billing due to need to provision more memory for the task?

**Comment by Fydon** [comment]
It would be helpful to know when the buffer has reached `max-buffer-size` no matter the mode, as in both cases there will be consequences and if it occurs frequently then `max-buffer-size` may need to be adjusted.

Are there plans to have metrics available to CloudWatch to enable alerting? If there was an OpenTelemetry option, that would be great but as it's possible to pull CloudWatch metrics to another solution CloudWatch should be the priority.

**Comment by AntonioDLuffy** [comment]
+1

Adding my support to this feature request. This feature would go a long way toward improving reliability and observability in containerized environments when using non-blocking mode.

**Comment by marvelph** [comment]
+1

Currently, it is difficult to find an appropriate value for max-buffer-size.

**Comment by choeflake** [comment]
> +1
> 
> Currently, it is difficult to find an appropriate value for max-buffer-size.

Indeed! And requirements could change over time. Now we have to re-evaluate the max-buffer-size regularly.

+1

**Comment by dipnile** [comment]
We are also currently facing difficulties determining the appropriate value for max-buffer-size. We also plan to continuously monitor the buffer to detect any potential log loss. Do you have any recommendations or best practices for this?
